### PR TITLE
Tests

### PR DESCRIPTION
**Pull request by avshiliaev**

### CODE REVIEW

File: .github/workflows/github-actions-tests.yml [hunk 1]
```diff
@@ -0,0 +1,18 @@
+on: [push, pull_request]
+name: Test
+jobs:
+  test:
+    strategy:
+      matrix:
+        go-version: [1.12.x]
```
Comment (aaannz) [inline .github/workflows/github-actions-tests.yml:7]: What is the reason to go with 1.12 version? With SLE15 we start on 1.14 up to 1.17. I suspect Leap will be the same.

Comment (avshiliaev) [reply .github/workflows/github-actions-tests.yml:7]: Well, it was set for the project in [go.mod](https://github.com/uyuni-project/inter-server-sync/blob/tests/go.mod#L3) but I think we can safely update the version, as far as I can see.

File: tests/utils.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+// MockWriter allows to create a mock bufferWriter object, as it implements the interface
+type MockWriter struct {
+	data []string
+}
+
+func (mr *MockWriter) Write(p []byte) (n int, err error) {
+
+	mr.data = append(mr.data, string(p))
+	return 4096, nil
```
Comment (rjmateus) [inline tests/utils.go:76]: shouldn't we return the length of p?

**Review by rjmateus — COMMENTED** [review]
Just some small comments.
Thanks for the great contribution

File: .github/workflows/github-actions-tests.yml (hunk 1 above)
Comment (avshiliaev) [reply .github/workflows/github-actions-tests.yml:7]: Got it, I'll update the version

File: dumper/utils_test.go [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+		if len(graph[node]) == 0 {
+			path = path[:len(path)-1]
+		}
+	}
+
+	return result
+}
+
+// setNumberOfRecordsForTable takes a generic mocked DataDumper object and simulates a case where a table X has Y records
+func setNumberOfRecordsForTable(tc *writerTestCase, tableName string, num int) {
```
[the remark this answers is not in the document]
Comment (avshiliaev) [reply dumper/utils_test.go:127]: @rjmateus  Exactly, it is being referred to in a testing function, which is commented out [here](setNumberOfRecordsForTable)

File: tests/utils.go (hunk 2 above)
Comment (avshiliaev) [reply tests/utils.go:76]: @rjmateus oh right, according to the documentation
> It returns the number of bytes written.
I thought it was the size of the buffer instead. Will change it

File: .github/workflows/github-actions-tests.yml (hunk 1 above)
Comment (aaannz) [reply .github/workflows/github-actions-tests.yml:7]: I rechecked and underlying SLE15SP3 has 1.17 already ( https://scc.suse.com/packages?name=SUSE%20Manager%20Server&version=4.2&arch=x86_64&query=go1.17&package_id=21579416 )
The same is true for SUMA 4.1 and SLE15SP2 ( https://scc.suse.com/packages?name=SUSE%20Manager%20Server&version=4.1&arch=x86_64&query=go1.17&package_id=21578693 )
So I see no reason to go with anything older then with 1.17.

Comment (rjmateus) [reply .github/workflows/github-actions-tests.yml:7]: Just one note, the version in use should be the one supported by OBS to build the project for those OS's (I think is the same one supported in the OS).
In run-time we don't need go.

File: dumper/utils_test.go [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+	// Two stacks approach
+	stack := []string{root}
+	var path []string
+
+	for len(stack) > 0 {
+		// pop the next node from the stack
+		node, stack = stack[len(stack)-1], stack[:len(stack)-1]
+
+		// update the current path if we're done with all paths from a branch
+		if len(stack) == 0 && node != root {
```
[the remark this answers is not in the document]
Comment (avshiliaev) [reply dumper/utils_test.go:101]: @rjmateus Thanks, that's some very cool point. If I get it right, and the tests kind of prove it, the problem boils down to finding all the valid paths in a tree. Well yes, the line here is needed, While we're traversing depth-first and our main stack runs out to 1 (before we pop), which means we searched the first sub-tree, we also need to reset the second stack, from where we read our result. However, it made me think of a cleaner and shorter solution. Is just like pre-order traversal?

Just to make it visual. This is an original graph:
```
{
    "root": [
        "v11",
        "v12"
    ],
    "v11": [],
    "v12": [
        "v13"
    ],
    "v13": [
        "v14"
    ],
    "v14": [
        "v15"
    ],
    "v15": [
        "v14"
    ]
}
```

and this is what we want to get from it:
```
{
    "root": true,
    "root,v11": true,
    "root,v12": true,
    "root,v12,v13": true,
    "root,v12,v13,v14": true,
    "root,v12,v13,v14,v15": true
}
```

File: dumper/utils_test.go [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+		dataDumper.TableData[node] = TableDump{
+			TableName: node,
+			KeyMap:    map[string]bool{fmt.Sprintf("'%04d'", 1): true},
+			Keys:      []TableKey{{Key: map[string]string{"id": fmt.Sprintf("'%04d'", 1)}}},
+		}
+		dataDumper.Paths[node] = true
+
+		// move forward, check if we're done with a level, update the parent if so
+		queue = append(queue, graph[node]...)
+		if levelSize == 0 {
```
[the remark this answers is not in the document]
Comment (avshiliaev) [reply dumper/utils_test.go:77]: @rjmateus we want to somehow keep the reference of a parent node for the current level to populate `ReferencedBy`. But I'll try it with this exmple :smiley:

[the remark this answers is not in the document]
Comment (avshiliaev) [reply dumper/utils_test.go]: @rjmateus I've fixed the logic

File: dumper/utils_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (avshiliaev) [reply dumper/utils_test.go]: @rjmateus I've fixed it here as well

**Review by rjmateus — APPROVED** [review]
LGTM.
Thanks for the great contribution

**Pull request merged**